### PR TITLE
Refactored dracut module

### PR DESCRIPTION
**Pull request by nmeum**
This is a relatively big refactor of the ZFS dracut module. It works perfectly for my setup, the question is: does it work for other people as well? I would encourage more people to this.

The most notable change is that `zfs-lib.sh` was added which provides utility functions used by `zfs-export.sh` and `zfs-mount.sh`. This change made the code much cleaner and more modular and I guess that this is the biggest achievement of this refactoring. Furthermore, it should be much easier now to add additional functionality to the dracut module.

I don't know if it was possible before but you can also use a pool instead of a dataset as your root by setting the bootfs attribute on the pool directly using: `zpool set bootfs=zroot zroot` (this should be documented in the `README` by the way). Besides it also fixes a few bugs here and there…

Basically there are two things left to do here:

- [x] Update the README
- [ ] Test this for different setups

### CODE REVIEW

**Comment by behlendorf** [comment]
@nmeum could you address the build failures, rebase this against master, and force push the branch for a new round of testing.

**Comment by nmeum** [comment]
I rebased this a few days ago. The built failures don't seem to be my fault…

**Comment by behlendorf** [comment]
Merged as:

ec1c11d Update dracut README
01d01ef Only source '/lib/dracut-lib.sh' if it wasn't so far
edbaf24 Force export if it failed the first time
eda3d4e Refactor dracut module

@nmeum these changes look good to me but I've only tested them lightly.  If you have a little time it would be great if you could cleanup the dracut code a little further.  Specifically, we'd like to get all the scripts to pass [shellcheck](http://www.shellcheck.net/) so we can start running it automatically in the automated testing.

**Comment by nmeum** [comment]
I am currently somewhat busy with things but If I get around to it I might cleanup the code a little further. With a few more changes it might as well be possible to get this included in the dracut upstream project...

**Comment by FransUrbo** [comment]
@nmeum Before you do, could you please have a look at #3560 first? That is what we call "Phase 3". Phase 1 was accepted last week and hopefully "Phase 2" (#3559) will be accepted after <code>0.6.5</code> is tagged/released.

With "Phase 3", I'm moving a lot of functionality from dracut to the common <code>zfs-functions.sh</code> script. Even if/when you want to start pushing dracut upstream, I still think it's a good idea to at least extract that code and include that in the upstream push.

I also noted that you remove some functionality from dracut while you where 'messing' with it. So I've added that back, AND some...
